### PR TITLE
[Compatibility] Added ZMPOP command

### PR DESCRIPTION
**Pull request by Vijay-Nirmal**
Adding the ZMPOP commands to garnet

- [x] Add ZMPOP commands
- [x] Add Integration Test cases, ACL Test and Slot Verification Test
- [x] Add documentation

### CODE REVIEW

File: libs/server/Storage/Session/ObjectStore/SortedSetOps.cs [hunk 1]
```diff
@@ -927,5 +927,58 @@ private GarnetStatus SortedSetDifference<TObjectContext>(ReadOnlySpan<ArgSlice>
...
+        /// </summary>
+        public unsafe GarnetStatus SortedSetMPop(ReadOnlySpan<ArgSlice> keys, int count, bool lowScoresFirst, out ArgSlice poppedKey, out (ArgSlice member, ArgSlice score)[] pairs)
+        {
+            pairs = default;
+            poppedKey = default;
+
+            if (keys.Length == 0)
+                return GarnetStatus.OK;
+
+            var createTransaction = false;
```
Comment (Vijay-Nirmal) [inline libs/server/Storage/Session/ObjectStore/SortedSetOps.cs:942]: I am not sure If I need to use transaction for this comment as the other similar command in Garnet doesn't have transaction

**Review by vazois — CHANGES_REQUESTED** [review]

File: libs/server/Resp/Objects/SortedSetCommands.cs [hunk 2]
```diff
@@ -356,6 +356,122 @@ private unsafe bool SortedSetPop<TGarnetApi>(RespCommand command, ref TGarnetApi
             return true;
         }
 
+        /// <summary>
+        /// Removes and returns up to count members from the first non-empty sorted set key from the list of keys.
+        /// </summary>
+        private unsafe bool SortedSetMPop<TGarnetApi>(ref TGarnetApi storageApi)
```
[the remark this answers is not in the document]
Comment (Vijay-Nirmal) [reply libs/server/Resp/Objects/SortedSetCommands.cs:362]: Interestingly, when I tested it, most of the existing comments that operate on multi-key had this issue. Below is an example of that. I have added the below code to check this. Let me know if we have to add it to all the existing multi-key obj commands, I can do it in a separate PR

```csharp
if (txnManager.ObjectStoreLockableContext.Session is null)
    ThrowObjectStoreUninitializedException();
```

image

[the remark this answers is not in the document]
Comment (vazois) [reply libs/server/Resp/Objects/SortedSetCommands.cs:363]: You can maybe do it for this command in this PR and if you have time to address the other commands it should be fine.

[the remark this answers is not in the document]
Comment (Vijay-Nirmal) [reply libs/server/Resp/Objects/SortedSetCommands.cs:363]: Already one for this PR

File: libs/server/Storage/Session/ObjectStore/SortedSetOps.cs (hunk 1 above)
Comment (TalZaccai) [reply libs/server/Storage/Session/ObjectStore/SortedSetOps.cs:945]: It depends if this operation should be atomic or not, I don't see any promise of atomicity in the spec. We don't treat LMPOP as a transaction either, I think that's probably on purpose.

File: libs/server/Resp/Objects/SortedSetCommands.cs [hunk 3]
```diff
@@ -356,6 +356,122 @@ private unsafe bool SortedSetPop<TGarnetApi>(RespCommand command, ref TGarnetApi
...
+            where TGarnetApi : IGarnetApi
+        {
+            // ZMPOP requires at least 3 args: numkeys, key [key...], <MIN|MAX> [COUNT count]
+            if (parseState.Count < 3)
+            {
+                return AbortWithWrongNumberOfArguments(nameof(RespCommand.ZMPOP));
+            }
+
+            // Get number of keys
+            if (!parseState.TryGetInt(0, out var numKeys) || numKeys < 1)
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SortedSetCommands.cs:373]: If `numKeys < 0` - `ERR numkeys should be greater than 0`

File: libs/server/Resp/Objects/SortedSetCommands.cs [hunk 4]
```diff
@@ -356,6 +356,122 @@ private unsafe bool SortedSetPop<TGarnetApi>(RespCommand command, ref TGarnetApi
...
+
+                var countArg = parseState.GetArgSliceByRef(numKeys + 2);
+                if (!countArg.ReadOnlySpan.EqualsUpperCaseSpanIgnoringCase(CmdStrings.COUNT))
+                {
+                    while (!RespWriteUtils.WriteError(CmdStrings.RESP_SYNTAX_ERROR, ref dcurr, dend))
+                        SendAndReset();
+                    return true;
+                }
+
+                if (!parseState.TryGetInt(numKeys + 3, out count) || count < 1)
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SortedSetCommands.cs:423]: If count < 1 - `ERR count should be greater than 0`

**Review by TalZaccai — COMMENTED** [review]
Looks good overall, couple of small comments

File: libs/server/Storage/Session/ObjectStore/SortedSetOps.cs (hunk 1 above)
Comment (Vijay-Nirmal) [reply libs/server/Storage/Session/ObjectStore/SortedSetOps.cs:945]: What should I do here? Personally I think we need transaction

File: libs/server/Resp/Objects/SortedSetCommands.cs [hunk 5]
```diff
@@ -356,6 +356,122 @@ private unsafe bool SortedSetPop<TGarnetApi>(RespCommand command, ref TGarnetApi
...
+                        // Write key
+                        while (!RespWriteUtils.WriteBulkString(poppedKey.ReadOnlySpan, ref dcurr, dend))
+                            SendAndReset();
+
+                        // Write array of member-score pairs
+                        while (!RespWriteUtils.WriteArrayLength(pairs.Length * 2, ref dcurr, dend))
+                            SendAndReset();
+
+                        foreach (var (member, score) in pairs)
+                        {
```
[the remark this answers is not in the document]
Comment (Vijay-Nirmal) [reply libs/server/Resp/Objects/SortedSetCommands.cs:458]: Any particular reason to use an array of arrays here? Personally, I like the structure provided by named tuple. Also the existing method (`SortedSetPop`) uses `(ArgSlice member, ArgSlice score)[]`

**Comment by Vijay-Nirmal** [comment]
@TalZaccai Fixed review command with an open question

File: libs/server/Storage/Session/ObjectStore/SortedSetOps.cs (hunk 1 above)
Comment (vazois) [reply libs/server/Storage/Session/ObjectStore/SortedSetOps.cs:945]: Please implement it as a txn.

File: libs/server/Resp/Objects/SortedSetCommands.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (vazois) [reply libs/server/Resp/Objects/SortedSetCommands.cs:457]: Are there any performance implications?

File: libs/server/Storage/Session/ObjectStore/SortedSetOps.cs (hunk 1 above)
Comment (Vijay-Nirmal) [reply libs/server/Storage/Session/ObjectStore/SortedSetOps.cs:945]: Then we are good right, I already have transaction as part of this PR

**Review by vazois — APPROVED** [review]
Looks good! Please resolve the remaining (if any) comments from @TalZaccai

File: libs/server/Resp/Objects/SortedSetCommands.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (TalZaccai) [reply libs/server/Resp/Objects/SortedSetCommands.cs:457]: I meant with regards to the RESP format, there should be an array length of 2 written before you write these two bulk strings, right?

[the remark this answers is not in the document]
Comment (Vijay-Nirmal) [reply libs/server/Resp/Objects/SortedSetCommands.cs:457]: Fixed

**Comment by Vijay-Nirmal** [comment]
@vazois Done

**Pull request merged**